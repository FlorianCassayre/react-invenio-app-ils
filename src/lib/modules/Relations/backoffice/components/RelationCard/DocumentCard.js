import { recordToPidType } from '@api/utils';
import { LiteratureCover } from '@modules/Literature/LiteratureCover';
import { DocumentAuthors } from '@modules/Document/DocumentAuthors';
import { DocumentTitle } from '@modules/Document/DocumentTitle';
import { BackOfficeRoutes } from '@routes/urls';
import _get from 'lodash/get';
import _isEmpty from 'lodash/isEmpty';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { Card, Icon } from 'semantic-ui-react';

export class DocumentCard extends Component {
  render() {
    const { data, extra, actions } = this.props;
    const linkTo = BackOfficeRoutes.documentDetailsFor(data.metadata.pid);
    return (
      <Card centered className="bo-relation-card" data-test={data.metadata.pid}>
        <Card.Meta className="discrete">
          {actions}
          {data.metadata.document_type || data.metadata.mode_of_issuance}
        </Card.Meta>
        {recordToPidType(data) === 'docid' ? (
          <LiteratureCover
            size="small"
            url={_get(data, 'metadata.cover_metadata.urls.medium')}
          />
        ) : (
          <Icon name="clone outline" size="huge" color="grey" />
        )}
        <Card.Content>
          <Card.Header as={Link} to={linkTo} target="_blank">
            <DocumentTitle metadata={data.metadata} truncate titleOnly />
          </Card.Header>
          <Card.Meta>
            <DocumentAuthors metadata={data.metadata} />
          </Card.Meta>
        </Card.Content>
        {!_isEmpty(extra) && <Card.Content extra>{extra}</Card.Content>}
      </Card>
    );
  }
}

DocumentCard.propTypes = {
  data: PropTypes.object.isRequired,
  extra: PropTypes.node,
  actions: PropTypes.node.isRequired,
};
DocumentCard.defaultProps = {
  extra: null,
};
