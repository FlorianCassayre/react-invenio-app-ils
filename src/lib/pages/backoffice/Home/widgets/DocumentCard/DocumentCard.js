import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Loader, Error } from '@components';
import { RecordsBriefCard } from '@components/backoffice/RecordsBriefCard';
import { NewButton, SeeAllButton } from '@components/backoffice/buttons';
import { BackOfficeRoutes } from '@routes/urls';
import { documentApi } from '@api';

export default class DocumentCard extends Component {
  constructor(props) {
    super(props);
    this.seeAllUrl = BackOfficeRoutes.documentsListWithQuery;
    this.newDocumentURL = BackOfficeRoutes.documentCreate;
  }

  componentDidMount() {
    const { fetchRequestedWithAvailableItems } = this.props;
    fetchRequestedWithAvailableItems();
  }

  seeAllButton = () => {
    const path = this.seeAllUrl(
      documentApi
        .query()
        .withAvailableItems()
        .withPendingLoans()
        .qs()
    );
    return <SeeAllButton fluid disabled to={path} />;
  };

  newDocumentButton = () => {
    return <NewButton fluid disabled to={this.newDocumentURL} />;
  };

  render() {
    const { data, isLoading, error } = this.props;
    return (
      <Loader isLoading={isLoading}>
        <Error error={error}>
          <RecordsBriefCard
            title={'Documents'}
            stats={data}
            text={'requested on shelf'}
            buttonLeft={this.newDocumentButton()}
            buttonRight={this.seeAllButton()}
          />
        </Error>
      </Loader>
    );
  }
}

DocumentCard.propTypes = {
  fetchRequestedWithAvailableItems: PropTypes.func.isRequired,
  data: PropTypes.number.isRequired,
  isLoading: PropTypes.bool.isRequired,
  error: PropTypes.object,
};

DocumentCard.defaultProps = {
  error: null,
};
