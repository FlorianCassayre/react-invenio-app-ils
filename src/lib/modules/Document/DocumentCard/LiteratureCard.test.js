import React from 'react';
import { mount } from 'enzyme';
import testData from '@testData/documents.json';
import DocumentCard from './DocumentCard';

describe('BookCard tests', () => {
  let component;
  afterEach(() => {
    if (component) {
      component.unmount();
    }
  });

  const bookData = {
    metadata: {
      ...testData[0],
      title: 'Lorem',
      edition: '12',
      authors: ['Author1', 'Author2'],
      imageSize: 'small',
      imageCover: '',
      circulation: { has_items_for_loan: 2 },
      eitems: { total: 2 },
    },
  };

  it('should render the BookCard', () => {
    component = mount(<DocumentCard data={bookData} />);
    expect(component).toMatchSnapshot();

    const rows = component
      .find('BookCard')
      .find('Card')
      .filterWhere(
        element => element.prop('data-test') === bookData.metadata.pid
      );
    expect(rows).toHaveLength(1);
  });
});
