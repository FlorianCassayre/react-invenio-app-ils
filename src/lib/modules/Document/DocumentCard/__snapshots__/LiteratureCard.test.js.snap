// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookCard tests should render the BookCard 1`] = `
<Overridable(DocumentCard)
  data={
    Object {
      "metadata": Object {
        "$schema": "https://127.0.0.1:5000/schemas/documents/document-v1.0.0.json",
        "abstract": "This is an abstract",
        "authors": Array [
          "Author1",
          "Author2",
        ],
        "circulation": Object {
          "has_items_for_loan": 2,
        },
        "document_type": "BOOK",
        "edition": "12",
        "eitems": Object {
          "total": 2,
        },
        "imageCover": "",
        "imageSize": "small",
        "keywords": Object {
          "source": "X",
          "value": "Patata",
        },
        "languages": Array [
          "en",
        ],
        "pid": "docid-1",
        "publication_year": "1950",
        "title": "Lorem",
      },
    }
  }
>
  <DocumentCard
    data={
      Object {
        "metadata": Object {
          "$schema": "https://127.0.0.1:5000/schemas/documents/document-v1.0.0.json",
          "abstract": "This is an abstract",
          "authors": Array [
            "Author1",
            "Author2",
          ],
          "circulation": Object {
            "has_items_for_loan": 2,
          },
          "document_type": "BOOK",
          "edition": "12",
          "eitems": Object {
            "total": 2,
          },
          "imageCover": "",
          "imageSize": "small",
          "keywords": Object {
            "source": "X",
            "value": "Patata",
          },
          "languages": Array [
            "en",
          ],
          "pid": "docid-1",
          "publication_year": "1950",
          "title": "Lorem",
        },
      }
    }
    volume={null}
  >
    <Overridable
      data={
        Object {
          "metadata": Object {
            "$schema": "https://127.0.0.1:5000/schemas/documents/document-v1.0.0.json",
            "abstract": "This is an abstract",
            "authors": Array [
              "Author1",
              "Author2",
            ],
            "circulation": Object {
              "has_items_for_loan": 2,
            },
            "document_type": "BOOK",
            "edition": "12",
            "eitems": Object {
              "total": 2,
            },
            "imageCover": "",
            "imageSize": "small",
            "keywords": Object {
              "source": "X",
              "value": "Patata",
            },
            "languages": Array [
              "en",
            ],
            "pid": "docid-1",
            "publication_year": "1950",
            "title": "Lorem",
          },
        }
      }
      id="DocumentCard.layout"
      volume={null}
    >
      <Card
        centered={true}
        className="fs-book-card"
        data-test="docid-1"
        link={true}
        onClick={[Function]}
      >
        <a
          className="ui centered link card fs-book-card"
          data-test="docid-1"
          onClick={[Function]}
        >
          <CardMeta
            className="discrete"
          >
            <div
              className="meta discrete"
            >
              BOOK
            </div>
          </CardMeta>
          <Overridable
            data={
              Object {
                "metadata": Object {
                  "$schema": "https://127.0.0.1:5000/schemas/documents/document-v1.0.0.json",
                  "abstract": "This is an abstract",
                  "authors": Array [
                    "Author1",
                    "Author2",
                  ],
                  "circulation": Object {
                    "has_items_for_loan": 2,
                  },
                  "document_type": "BOOK",
                  "edition": "12",
                  "eitems": Object {
                    "total": 2,
                  },
                  "imageCover": "",
                  "imageSize": "small",
                  "keywords": Object {
                    "source": "X",
                    "value": "Patata",
                  },
                  "languages": Array [
                    "en",
                  ],
                  "pid": "docid-1",
                  "publication_year": "1950",
                  "title": "Lorem",
                },
              }
            }
            id="DocumentCard.cover"
            volume={null}
          >
            <Overridable(LiteratureCover)
              size="small"
            >
              <LiteratureCover
                asItem={false}
                isRestricted={false}
                linkTo={null}
                size="small"
                url={null}
              >
                <Overridable
                  asItem={false}
                  id="LiteratureCover.placeholder"
                  isRestricted={false}
                  linkTo={null}
                  size="small"
                  url={null}
                >
                  <Placeholder>
                    <div
                      className="ui placeholder"
                    >
                      <PlaceholderImage
                        square={true}
                      >
                        <div
                          className="square image"
                        />
                      </PlaceholderImage>
                    </div>
                  </Placeholder>
                </Overridable>
              </LiteratureCover>
            </Overridable(LiteratureCover)>
          </Overridable>
          <CardContent>
            <div
              className="content"
            >
              <CardHeader>
                <div
                  className="header"
                >
                  Lorem
                </div>
              </CardHeader>
              <CardMeta>
                <div
                  className="meta"
                >
                  <Overridable(DocumentAuthors)
                    authorsLimit={10}
                    metadata={
                      Object {
                        "$schema": "https://127.0.0.1:5000/schemas/documents/document-v1.0.0.json",
                        "abstract": "This is an abstract",
                        "authors": Array [
                          "Author1",
                          "Author2",
                        ],
                        "circulation": Object {
                          "has_items_for_loan": 2,
                        },
                        "document_type": "BOOK",
                        "edition": "12",
                        "eitems": Object {
                          "total": 2,
                        },
                        "imageCover": "",
                        "imageSize": "small",
                        "keywords": Object {
                          "source": "X",
                          "value": "Patata",
                        },
                        "languages": Array [
                          "en",
                        ],
                        "pid": "docid-1",
                        "publication_year": "1950",
                        "title": "Lorem",
                      }
                    }
                  >
                    <DocumentAuthors
                      allFields={false}
                      authorsLimit={10}
                      delimiter="; "
                      listItemAs={null}
                      metadata={
                        Object {
                          "$schema": "https://127.0.0.1:5000/schemas/documents/document-v1.0.0.json",
                          "abstract": "This is an abstract",
                          "authors": Array [
                            "Author1",
                            "Author2",
                          ],
                          "circulation": Object {
                            "has_items_for_loan": 2,
                          },
                          "document_type": "BOOK",
                          "edition": "12",
                          "eitems": Object {
                            "total": 2,
                          },
                          "imageCover": "",
                          "imageSize": "small",
                          "keywords": Object {
                            "source": "X",
                            "value": "Patata",
                          },
                          "languages": Array [
                            "en",
                          ],
                          "pid": "docid-1",
                          "publication_year": "1950",
                          "title": "Lorem",
                        }
                      }
                      otherAuthorsDisplay={null}
                      prefix=""
                      scrollLimit={Infinity}
                    >
                      <Overridable
                        allFields={false}
                        authorsLimit={10}
                        delimiter="; "
                        id="DocumentAuthors.layout"
                        listItemAs={null}
                        metadata={
                          Object {
                            "$schema": "https://127.0.0.1:5000/schemas/documents/document-v1.0.0.json",
                            "abstract": "This is an abstract",
                            "authors": Array [
                              "Author1",
                              "Author2",
                            ],
                            "circulation": Object {
                              "has_items_for_loan": 2,
                            },
                            "document_type": "BOOK",
                            "edition": "12",
                            "eitems": Object {
                              "total": 2,
                            },
                            "imageCover": "",
                            "imageSize": "small",
                            "keywords": Object {
                              "source": "X",
                              "value": "Patata",
                            },
                            "languages": Array [
                              "en",
                            ],
                            "pid": "docid-1",
                            "publication_year": "1950",
                            "title": "Lorem",
                          }
                        }
                        otherAuthorsDisplay={null}
                        prefix=""
                        scrollLimit={Infinity}
                      >
                        <div
                          className="document-authors-list-wrapper "
                        >
                          <List
                            as={null}
                            className="document-authors-list"
                            horizontal={true}
                          >
                            <div
                              className="ui horizontal list document-authors-list"
                              role="list"
                            >
                              <ListItem
                                key="Key0"
                              >
                                <div
                                  className="item"
                                  onClick={[Function]}
                                  role="listitem"
                                >
                                  ; 
                                </div>
                              </ListItem>
                              <ListItem
                                key="Key1"
                              >
                                <div
                                  className="item"
                                  onClick={[Function]}
                                  role="listitem"
                                />
                              </ListItem>
                            </div>
                          </List>
                        </div>
                      </Overridable>
                    </DocumentAuthors>
                  </Overridable(DocumentAuthors)>
                  <div>
                    Edition 
                    12
                  </div>
                </div>
              </CardMeta>
            </div>
          </CardContent>
          <CardContent
            extra={true}
          >
            <div
              className="extra content"
            >
              <Label>
                <div
                  className="ui label"
                  onClick={[Function]}
                >
                  On shelf
                </div>
              </Label>
              <Label>
                <div
                  className="ui label"
                  onClick={[Function]}
                >
                  E-book
                </div>
              </Label>
            </div>
          </CardContent>
        </a>
      </Card>
    </Overridable>
  </DocumentCard>
</Overridable(DocumentCard)>
`;
